cmake/introspection.cmake

```diff
@@ -17,3 +19,215 @@ CHECK_INCLUDE_FILE_CXX(sys/select.h HAVE_SYS_SELECT_H)
+  check_cxx_symbol_exists(be16toh ${ENDIAN_HEADER} HAVE_DECL_BE16TOH)
+  check_cxx_symbol_exists(be32toh ${ENDIAN_HEADER} HAVE_DECL_BE32TOH)
+  check_cxx_symbol_exists(be64toh ${ENDIAN_HEADER} HAVE_DECL_BE64TOH)
+  check_cxx_symbol_exists(htobe16 ${ENDIAN_HEADER} HAVE_DECL_HTOBE16)
+  check_cxx_symbol_exists(htobe32 ${ENDIAN_HEADER} HAVE_DECL_HTOBE32)
+  check_cxx_symbol_exists(htobe64 ${ENDIAN_HEADER} HAVE_DECL_HTOBE64)
+  check_cxx_symbol_exists(htole16 ${ENDIAN_HEADER} HAVE_DECL_HTOLE16)
+  check_cxx_symbol_exists(htole32 ${ENDIAN_HEADER} HAVE_DECL_HTOLE32)
+  check_cxx_symbol_exists(htole64 ${ENDIAN_HEADER} HAVE_DECL_HTOLE64)
+  check_cxx_symbol_exists(le16toh ${ENDIAN_HEADER} HAVE_DECL_LE16TOH)
+  check_cxx_symbol_exists(le32toh ${ENDIAN_HEADER} HAVE_DECL_LE32TOH)
+  check_cxx_symbol_exists(le64toh ${ENDIAN_HEADER} HAVE_DECL_LE64TOH)
+endif()
+
+CHECK_INCLUDE_FILE_CXX(unistd.h HAVE_UNISTD_H)
+if(HAVE_UNISTD_H)
+  check_cxx_symbol_exists(fdatasync "unistd.h" HAVE_FDATASYNC)
+  check_cxx_symbol_exists(fork "unistd.h" HAVE_DECL_FORK)
+  check_cxx_symbol_exists(pipe2 "unistd.h" HAVE_DECL_PIPE2)
+  check_cxx_symbol_exists(setsid "unistd.h" HAVE_DECL_SETSID)
+endif()
+
+CHECK_INCLUDE_FILE_CXX(sys/types.h HAVE_SYS_TYPES_H)
+CHECK_INCLUDE_FILE_CXX(ifaddrs.h HAVE_IFADDRS_H)
+if(HAVE_SYS_TYPES_H AND HAVE_IFADDRS_H)
+  check_cxx_symbol_exists(freeifaddrs "sys/types.h;ifaddrs.h" HAVE_DECL_FREEIFADDRS)
+  check_cxx_symbol_exists(getifaddrs "sys/types.h;ifaddrs.h" HAVE_DECL_GETIFADDRS)
+  # Illumos/SmartOS requires linking with -lsocket if
+  # using getifaddrs & freeifaddrs.
+  # See: https://github.com/bitcoin/bitcoin/pull/21486
+  if(HAVE_DECL_GETIFADDRS AND HAVE_DECL_FREEIFADDRS)
+    set(check_socket_source "
+      #include <sys/types.h>
+      #include <ifaddrs.h>
+
+      int main() {
+        struct ifaddrs* ifaddr;
+        getifaddrs(&ifaddr);
+        freeifaddrs(ifaddr);
+      }
+    ")
+    check_cxx_source_compiles("${check_socket_source}" IFADDR_WORKS_WITHOUT_LINK)
+    if(NOT IFADDR_WORKS_WITHOUT_LINK)
+      set(CMAKE_REQUIRED_LIBRARIES socket)
+      check_cxx_source_compiles("${check_socket_source}" IFADDR_NEEDS_LINK_TO_SOCKET)
+      set(CMAKE_REQUIRED_LIBRARIES)
+      if(IFADDR_NEEDS_LINK_TO_SOCKET)
+        link_libraries(socket)
+      else()
+        message(FATAL_ERROR "Cannot figure out how to use getifaddrs/freeifaddrs.")
+      endif()
+    endif()
+  endif()
+endif()
+
+# Check for gmtime_r(), fallback to gmtime_s() if that is unavailable.
+# Fail if neither are available.
+check_cxx_source_compiles("
+#include <ctime>
+
+int main()
+{
+  gmtime_r((const time_t*)nullptr, (struct tm*)nullptr);
+}
+" HAVE_GMTIME_R)
+if(NOT HAVE_GMTIME_R)
+  check_cxx_source_compiles("
+  #include <ctime>
+
+  int main()
+  {
+    gmtime_s((struct tm*)nullptr, (const time_t*)nullptr);
+  }
+  " HAVE_GMTIME_S)
+  if(NOT HAVE_GMTIME_S)
+    message(FATAL_ERROR "Both gmtime_r and gmtime_s are unavailable.")
+  endif()
+endif()
+
+if(WIN32)
```

Comment:
I'd rather not start special-casing things like this. They tend to get outdated very quickly that way. All we need to know here is "`is std::system` or `::_wsystem` available?". We don't need to worry about which platforms to tie those checks to.

Doing it that way would continue to work in a future where, for example, `wsystem` is removed in windows and replaced with `std::system`. In that case, the check here would still pass as it should.

Reply:
Thanks! [Updated](https://github.com/hebasto/bitcoin/pull/6#issuecomment-1435777703).